### ISSUE TITLE
Proposal: VariablesOnTop

### ISSUE THREAD

**Issue by rrrene** [rrrene/credo#249]
We could have a check that encourages declaring all variables at the top of a function. If you are declaring several variables after your function already "did some work", your function might be too complex.

```elixir
# this can be difficult to comprehend ...

def update(id, params) do
  user = User.load(...)

  Endpoit.broadcast!(...)

  now = DateTime.now

  User.activate!(user, now)

  assigns = [user: user, now: now]
  render_response(params, assigns)
end

# ... in comparison with this:

def update(id, params) do
  user = User.load(...)
  now = DateTime.now
  assigns = [user: user, now: now]

  Endpoit.broadcast!(...)
  User.activate!(user, now)
  render_response(params, assigns)
end

# ... or this:

def update(id, params) do
  {user, now} = broadcast_and_activate(id, params)

  render_response(params, user: user, now: now)
end

def broadcast_and_activate(id, params) do
  user = User.load(...)
  now = DateTime.now

  Endpoit.broadcast!(...)
  User.activate!(user, now)

  {user, now}
end

```

This is of course a contrived example, but you get the idea. Each function should be as concise as possible.

**Comment by h4cc** [comment]
I think "encourage" is important to understand here. Having new variables declared late in a function can be considered a "bad smell" and could be solved using smaller functions.

**Comment by devonestes** [comment]
Hmmmm... In theory I like this since it encourages smaller functions and chunking of concepts, which is usually a good thing for reading comprehension, but I'm also a fan of using descriptive variable names at certain points in code to illustrate to the reader certain concepts. I know "Introduce Explaining Variable" is technically an OO refactoring concept, but I think it's one that works in just about any language if you have complex intermediate states that sometimes need a little more info to make sense to a reader. I could go either way on this one.

**Comment by erikmueller** [comment]
I'd like to try :)

**Comment by itsgreggreg** [comment]
Elixir/Erlang doesn't have variable hoisting, so this isn't really a code smell thing here. It's all just a matter of thought-process readability which IMHO credo can't check.
